### ISSUE TITLE
Ticket Contents and Design

### ISSUE THREAD

**Issue by ASAAR** [ASAAR/SE2-KaggleComp#104]
# Overview

So we know that in the long-term, we'll want a "ticket" that can be passed between the Hub and a Worker which contains all of the information needed for training and testing. The question is, what goes into this exactly and how do we handle it? I'll give a brief overview of my current direction and issues we need to handle.
# Key Questions
- Do we want a ticket to also contain "null" or empty data that will be filled with information that's needed on the return trip to the Hub? This brings up some issues surrounding responsibility and we might want to consider two objects, one that goes to the hub and one that comes back.
- Do we allow a worker to make multiple passes at the same work ticket? On the one hand, it would be an optimization because the worker could make slight changes and try to optimize results. On the other hand it would convolute the distinction between Hub and Worker.
- Should we use the name of the function for the model we are using? That brings up issues of backwards compatibility if we change the name of the function providing a model.
# Current Requirements
- Model Function Name: The name of the function that will be called to retrieve a model.
- Model Target (or "Label"): The column to be predicted.
- Training Data: The labeled data that is used for training.
- Testing Data: The unlabeled data that the trained model is used on.
- Hyperparameters to Use: If we are telling the system to get the best results possible, this is a good idea to have.
- Validation Results: Results of splitting the training data and running against it.
- Test Results: Results of running the model against the test data. Predictions on unlabeled data.
- Hyperparameters Used: We want to track which hyperparameters have been used.
- Model Used: The name of the function.
- Max Iterations: The maximum number of times that a worker can run against a dataset using the same model and making slight changes.
- Past Results: Past results for this model to help the system decide what to do in the next iteration.

Note: Although I have two boolean values for if the testing/training data is in a folder, those will be removed. We should setup auto-detection instead.

Am I missing anything? **Please tell me if I am**

**Comment by ZakeryFyke** [comment]
Regarding your first question, what all would we want to send back to the hub? Just what is listed under dependent on key question 1?
If we create a second object for the return trip to the Hub, would we still be able to make the multiple passes/tweaks that your second question suggests?

**Comment by ASAAR** [comment]
@ZakeryFyke It isn't an exhaustive list. I guess I'm looking for ideas on what could be added to it. 

Also, yes we could. Because it would do iterations and then send the results in chunks. Maybe?

**Comment by asclines** [comment]
I think it would be a good idea to have one ticket that gets modified on the way. Sort of like a record or current state of whats going on. That way we never have to worry about having old data (such as previous hyper parameters) and we can easily save the data for future use as its in one spot.

**Comment by asclines** [comment]
Also, @ASAAR can you clarify what you mean by a worker making multiple passes?

**Comment by ASAAR** [comment]
@asclines Sounds good! My only concern with regard to using function name is the MLTA integration. If we save that model as the function name, then we could have situations where results for a model are disconnected from prior results due to a name change.

A worker making multiple passes is like a worker training and testing multiple times with the same model but multiple hyper-parameters or data configurations. It would save some work, but cause complexity issues.

**Comment by asclines** [comment]
@ASAAR Well maybe we don't use function names but abstract out a layer and just have labels that are user defined? Or give each function a UID that never changes? <- This one I don't like as much

But on another note, how likely is there to be a situation where the name gets changed but nothing else? What I'm getting at is, maybe we don't want the results to be connected..


And yes, the worker should be able to make multiple passes with the same ticket. Somewhere on the ticket it should say something like max_iterations for how many times should the worker try and optimize parameters before giving up and declaring the model a failure.

**Comment by ASAAR** [comment]
@asclines Yeah, I agree that the situation is unlikely to matter. I just wanted to bring it up so that we aren't blindsided later. I think for now we can stick with our current system. If necessary later, we can add something to model_properties.py (or something like that) which acts as a uniqueID for MLTA.

Sounds good, I'll add that to the list.

**Comment by asclines** [comment]
Proposed hyper-parameter configuration: (From 10/30 meeting)

Hyper parameters should be handled on a model per model bases.
Default values and ranges per parameter. 
This only needs to be done when the user wants to override the model default options.
Will be done in configuration xml file

**Comment by ASAAR** [comment]
@asclines or anyone else, do we want to create classes that package parts of the ticket into input and output? Because otherwise the creation of the ticket gets ugly. The None values are temps, since we don't have a good way of doing it currently. However, you see the problem.

`class Ticket(object):

    def __init__(self, algorithm, train, test, label):
        self.ml_algorithm = algorithm
        self.training = train
        self.testing = test
        self.target = label
        self.hyperparameters = None
        self.validation_results = None
        self.test_results = None
        self.parameters_used = None
        self.max_iterations = 0
        self.past_results = None`

**Comment by asclines** [comment]
@ASAAR Could you clarify what you mean by packaging into input/output?

**Comment by ASAAR** [comment]
For example...

### Input
- ml_algorithm (from Hub)
- training (from config)
- testing (from config)
- target (from config)
- hyperparameters (retrieved from config)
- past_results (from MLTA)
- max_iterations (from config)

### Output
- validation_results
- test_results
- parameters_used

That way, it provides a clear separation and we could do something like this.

### Ticket
- ticket_input
- ticket_output

Make sense?

**Comment by asclines** [comment]
@ASAAR Would it make more sense to divide the variables under ticket by whether or not they are modifiable by the ticket? (Not 100% sure if I just reworded what you said or not.

**Comment by ASAAR** [comment]
Interesting.....modifiable by the ticket or the worker?

**Comment by asclines** [comment]
Are we gonna let the ticket be more than just a class that holds data?

**Comment by ASAAR** [comment]
I didn't think so, which is why I asked. Alright, yeah, I think that makes sense. It should end up as being the exact same separation though.
